pkg/dns/dns.go

```diff
@@ -268,6 +301,124 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
 	kd.newService(newObj)
 }
 
+func assertIsIngress(obj interface{}) (*ext.Ingress, bool) {
+	if ing, ok := obj.(*ext.Ingress); ok {
+		return ing, ok
+	} else {
+		glog.Errorf("Type assertion failed! Expected object of type \"Ingress\", got %T", obj)
+		return nil, ok
+	}
+}
+
+// newIngress generates DNS records for an Ingress resource.
+//
+// If all the Ingress loadbalancer status entries have their IP fields
+// set we generate a series of A records for them, one per IP. If there
+// is only one hostname in the Ingress status entries, we generate a
+// CNAME record for it. If there are multiple hostnames and no IP then
+// we generate a CNAME for the first hostname. If both IPs and hostnames
+// coexist, we favor IPs over the hostnames and we generate only A
+// records for the IPs and no CNAME record at all.
```

Comment:
An Ingress implementation could provide both a set of IP address endpoints and automatically create a DNS entry for it and put them down in the status fields. I don't think we have any implementation today that use these fields in this way, but it is entirely possible.